internal/pkg/deploy/cloudformation/app.go

```diff
@@ -7,35 +7,366 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/archer"
 	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/aws/cloudformation"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/aws/cloudformation/stackset"
 	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/deploy"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/deploy/cloudformation/stack"
+	sdkcloudformation "github.com/aws/aws-sdk-go/service/cloudformation"
+	sdkcloudformationiface "github.com/aws/aws-sdk-go/service/cloudformation/cloudformationiface"
 )
 
-// DeployApp deploys an application stack and waits until the deployment is done.
-// If the application doesn't exist, then it creates the stack.
-// If the application already exists, it updates the stack.
-func (cf CloudFormation) DeployApp(conf StackConfiguration, opts ...cloudformation.StackOption) error {
-	stack, err := toStack(conf)
+// DeployApp sets up everything required for our application-wide resources.
+// These resources include things that are regional, rather than scoped to a particular
+// environment, such as ECR Repos, CodePipeline KMS keys & S3 buckets.
+// We deploy application resources through StackSets - that way we can have one
+// template that we update and all regional stacks are updated.
+func (cf CloudFormation) DeployApp(in *deploy.CreateAppInput) error {
+	appConfig := stack.NewAppStackConfig(in)
+	s, err := toStack(appConfig)
 	if err != nil {
 		return err
 	}
-	for _, opt := range opts {
-		opt(stack)
+	if err := cf.cfnClient.CreateAndWait(s); err != nil {
+		// If the stack already exists - we can move on to creating the StackSet.
+		var alreadyExists *cloudformation.ErrStackAlreadyExists
+		if !errors.As(err, &alreadyExists) {
+			return err
+		}
 	}
 
-	err = cf.cfnClient.CreateAndWait(stack)
-	if err == nil { // Created a new stack, stop execution.
+	blankAppTemplate, err := appConfig.ResourceTemplate(&stack.AppResourcesConfig{
+		App: appConfig.Name,
+	})
+	if err != nil {
+		return err
+	}
+
+	return cf.appStackSet.Create(appConfig.StackSetName(), blankAppTemplate,
+		stackset.WithDescription(appConfig.StackSetDescription()),
+		stackset.WithExecutionRoleName(appConfig.StackSetExecutionRoleName()),
+		stackset.WithAdministrationRoleARN(appConfig.StackSetAdminRoleARN()),
+		stackset.WithTags(toMap(appConfig.Tags())))
+}
+
+// DelegateDNSPermissions grants the provided account ID the ability to write to this application's
+// DNS HostedZone. This allows us to perform cross account DNS delegation.
+func (cf CloudFormation) DelegateDNSPermissions(app *archer.Project, accountID string) error {
```

Comment:
When do you want to convert this over to `config`? since method is in the `deploy` pkg I wonder if this is a good time to do it. An alternative is to handle the archer pkg completely at the end.

Reply:
ok I'll try to get rid of this :+1:

Reply:
In the next PR